arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt

```diff
@@ -37,54 +36,50 @@ public fun <K, A, B> Map<K, A>.zip(other: Map<K, B>): Map<K, Pair<A, B>> =
  * and combining the elements with the given function.
  *
  * ```kotlin
- * import arrow.core.*
+ * import arrow.core.zip
+ * import io.kotest.matchers.shouldBe
  *
- * fun main(args: Array<String>) {
- *   //sampleStart
- *   val result =
- *    mapOf(1 to "A", 2 to "B").zip(mapOf(1 to "1", 2 to "2", 3 to "3")) {
- *      key, a, b -> "$key -> $a # $b"
- *    }
- *   //sampleEnd
- *   println(result)
+ * fun test() {
+ *   mapOf(1 to "A", 2 to "B").zip(mapOf(1 to "1", 2 to "2", 3 to "3")) {
+ *     key, a, b -> "$a ~ $b"
+ *   } shouldBe mapOf(1 to "A ~ 1", 2 to "B ~ 2")
  * }
  * ```
  * <!--- KNIT example-map-02.kt -->
+ * <!--- lines.isEmpty() -->
  */
-public inline fun <Key, A, B, C> Map<Key, A>.zip(other: Map<Key, B>, map: (Key, A, B) -> C): Map<Key, C> {
-  val destination = LinkedHashMap<Key, C>(size)
-  for ((key, bb) in this) {
-    Nullable.zip(other[key]) { cc -> map(key, bb, cc) }
-      ?.let { l -> destination.put(key, l) }
+public inline fun <Key, A, B, C> Map<Key, A>.zip(other: Map<Key, B>, map: (Key, A, B) -> C): Map<Key, C> =
+  buildMap(size) {
+    this@zip.forEach { (key, bb) ->
+      nullable {
```

Comment:
This is an interesting use of `nullable` to essentially not do this if the key is not present in any of the maps. Took me a while to understand what was going on...

Reply:
I updated this code in preparation of `Nullable.zip` deprecation.